### PR TITLE
added game loop, code comparison logic, win condition

### PR DESCRIPTION
**Pull request by lpenzey**
The program can now:
-quit at the beginning or enter their first guess
-loop through 10 user guesses
-give feedback about the number of red and white pins
-tell the user if they won
-allow the user to quit or restart if they won
-tells the user what the code was if unsuccessful after 10 tries

### CODE REVIEW

File: Lib/mastermind.rb [hunk 1]
```diff
@@ -1,24 +1,97 @@
...
-  def show_instructions
-    puts 'Welcome to Mastermind! Your challenge is
-to guess the secret four digit code in under 10 tries.
 
-The code can include the following colors:
+class Welcome
+  def show_instructions
+    puts "Welcome to Mastermind! Your challenge is
+to guess the secret four digit code in under 10 tries."
+    #sleep 5
```
Comment (ssunday) [inline Lib/mastermind.rb]: Removing commented out code is a good cleanup practice as it can be confusing

File: Lib/mastermind.rb [hunk 2]
```diff
@@ -1,24 +1,97 @@
...
+    abort("It's been an honor serving with you. Farewell!")
   end
 end
 
 class Code
   attr_reader :answer
+  attr_reader :player_input
+  attr_reader :red_pins
+  attr_reader :white_pins
+  attr_reader :position
```
Comment (ssunday) [inline Lib/mastermind.rb]: these can be put onto the same line or at least grouped accordingly

`attr_reader :red_pins, :white_pins`

File: Lib/mastermind.rb [hunk 3]
```diff
@@ -1,24 +1,97 @@
...
+    @answer = (1..4).map { @colors[rand(6)] }
+  end
+
+  def red_counter(key, input)
+    red_pins = []
+    position = 0
+    while input.length > position
+      key.each do |match|
+        red_pins << 1 if key[position] == input[position]
+        position += 1
```
Comment (ssunday) [inline Lib/mastermind.rb]: Actually, position should be removed from attr_reader as it's defined in every function that uses it, no? So it's scoped to each method

Comment (ssunday) [reply Lib/mastermind.rb]: I mean, between red_count/trim...unless they need to be shared?

File: Spec/mastermind_spec.rb [hunk 4]
```diff
@@ -23,6 +23,100 @@
...
+    expect(code.red_counter(key, input)).to eq [1, 1]
+  end
+end
+
+RSpec.describe Code, "trims code" do
+  let(:code) { Code.new }
+  let(:key) { ["r", "g", "y", "r"] }
+  let(:input) { ["r", "o", "b", "r"] }
+
+  it "deletes input position if it is the same as key index" do
```
Comment (ssunday) [inline Spec/mastermind_spec.rb]: For the same class described put all `it` cases within it and then have `contexts` or `describes` within it to give more detail on what it is doing. Keeps the specs tidy

File: Spec/mastermind_spec.rb [hunk 5]
```diff
@@ -23,6 +23,100 @@
...
+    expect(code.trim(key, input)).to eq ["", "", "", "r"]
+  end
+end
+
+RSpec.describe Code, "counts white pins" do
+  let(:code) { Code.new }
+  let(:key) { ["y", "b", "y", "r"] }
+  let(:input) { ["", "y", "b", ""] }
+
+  it "accurately counts white pins" do
```
Comment (ssunday) [inline Spec/mastermind_spec.rb]: Like `describe #white_counter` would be the block within `RSpec.describe Code` that would have all tests related to that function`

File: Spec/mastermind_spec.rb [hunk 6]
```diff
@@ -46,3 +140,10 @@
     expect { response.color_error }.to output(/Invalid colors/).to_stdout
   end
 end
+
+# ~> LoadError
+# ~> cannot load such file -- mastermind
+# ~>
+# ~> /Users/lucaspenzeymoog/.rbenv/versions/2.3.1/lib/ruby/2.3.0/rubygems/core_ext/kernel_require.rb:55:in `require'
```
Comment (ssunday) [inline Spec/mastermind_spec.rb]: 😮

File: Lib/mastermind.rb [hunk 7]
```diff
@@ -1,24 +1,97 @@
...
+    if input[position] == key[position]
+      input[position] = ""
+      key[position] = ""
+    end
+    position += 1
+    input
+  end
+
+  def white_counter(key, input)
+    x = (key & input).flat_map { |n| [n]*[key.count(n), input.count(n)].min }
```
Comment (pwentz) [inline Lib/mastermind.rb]: I would try to steer as far away from variable names like `x` and `y`, and other esoteric letters as much as possible. The only instance where they are useful is when you're doing something with a very generic argument (like `n` to represent a number). Stuff like this is low-hanging fruit for improving the readability of your code

File: Lib/mastermind.rb [hunk 8]
```diff
@@ -1,24 +1,97 @@
...
+    white_pins = x.reject { |c| c.empty? }
```
Comment (pwentz) [inline Lib/mastermind.rb]: You don't need to save `x.reject { |c| c.empty? }` as a local variable if you're not using it. You can just put the `reject` expression as the last line of the method.

File: Lib/mastermind.rb [hunk 9]
```diff
@@ -81,7 +72,7 @@ def white_counter(key, input)
   end
 
   def feedback(white_pins, red_pins)
-    puts "You have " + red_pins.length.to_s + " red pins and " + white_pins.compact.length.to_s + " white pins. Keep going!"
+    puts "You have " + red_pins.length.to_s + " red pins and " + white_pins.length.to_s + " white pins. Keep going!"
```
Comment (ssunday) [inline Lib/mastermind.rb]: "You have #{red_pins.length}", etc, is a little bit easier way to do this

**Pull request merged**